Stratis.Bitcoin/Features/BlockStore/LoopSteps/ProcessPendingStorageStep.cs

```diff
@@ -44,70 +44,75 @@ internal ProcessPendingStorageStep(BlockStoreLoop blockStoreLoop, ILoggerFactory
         {
             this.logger.LogTrace("({0}:'{1}/{2}',{3}:{4})", nameof(nextChainedBlock), nextChainedBlock.HashBlock, nextChainedBlock.Height, nameof(disposeMode), disposeMode);
 
-            BlockPair pendingBlockPairToStore;
-            if (!this.BlockStoreLoop.PendingStorage.TryRemove(nextChainedBlock.HashBlock, out pendingBlockPairToStore))
-                return StepResult.Next;
+            var context = new ProcessPendingStorageContext(this.BlockStoreLoop, nextChainedBlock);
 
-            var pendingBlockPairsToStore = new List<BlockPair>();
-            pendingBlockPairsToStore.Add(pendingBlockPairToStore);
-            int pendingStorageBatchSize = pendingBlockPairToStore.Block.GetSerializedSize();
+            //Next block does not exist in pending storage, continue onto the download blocks step.
+            if (!this.BlockStoreLoop.PendingStorage.ContainsKey(context.NextChainedBlock.HashBlock))
+                return StepResult.Next;
 
-            ChainedBlock lastFoundChainedBlock = nextChainedBlock;
+            //If pending storage has not yet reached the threshold or the node is in IDB, stop the loop and wait for more blocks.
+            if (this.BlockStoreLoop.PendingStorage.Count < BlockStoreLoop.PendingStorageBatchThreshold && this.BlockStoreLoop.ChainState.IsInitialBlockDownload == true)
```

Comment:
You can swap this so first check for IBD, to avoid enumerating the list.
Also need to take the dispose property in to account(i.e. if in dispose more then just push).